File: wallhaven_explorer_flutter/lib/core/services/image_cache_service.dart
// // AkonDeV 06/2026

import 'dart:io';
import 'package:dio/dio.dart';
import 'package:path/path.dart' as p;
import 'package:path_provider/path_provider.dart';

class ImageCacheService {
  final Dio _dio;

  ImageCacheService(this._dio);

  Future<String> getCachedImagePath(String id, String url) async {
    // // AkonDeV 06/2026
    final appDir = await getApplicationSupportDirectory();
    final cacheFolder = p.join(appDir.path, 'Cache');

    final cacheDir = Directory(cacheFolder);
    if (!await cacheDir.exists()) {
      await cacheDir.create(recursive: true);
    }

    String extension = p.extension(url);
    if (extension.isEmpty) extension = '.jpg';
    final localPath = p.join(cacheFolder, '$id$extension');

    final file = File(localPath);
    if (await file.exists()) {
      return localPath;
    }

    try {
      await _dio.download(url, localPath);
      return localPath;
    } catch (e) {
      throw HttpException('Fallo al descargar en caché local: $e');
    }
  }

  Future<void> clearCache() async {
    // // AkonDeV 06/2026
    final appDir = await getApplicationSupportDirectory();
    final cacheFolder = p.join(appDir.path, 'Cache');
    final cacheDir = Directory(cacheFolder);
    if (await cacheDir.exists()) {
      await cacheDir.delete(recursive: true);
    }
  }
}

File: WallhavenExplorer.Desktop/Services/ImageCacheService.cs
using System;
using System.IO;
using System.Net.Http;
using System.Threading.Tasks;
using WallhavenExplorer.Core.Services;

namespace WallhavenExplorer.Desktop.Services
{
    public class ImageCacheService : IImageCacheService
    {
        private readonly string _cacheFolder;
        private readonly IHttpClientFactory _clientFactory;

        public ImageCacheService(IHttpClientFactory clientFactory)
        {
            // // AkonDeV 06/2026
            _clientFactory = clientFactory;
            string appData = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
            _cacheFolder = Path.Combine(appData, "WallhavenExplorer", "Cache");
            if (!Directory.Exists(_cacheFolder)) Directory.CreateDirectory(_cacheFolder);
        }

        public async Task<string> GetCachedImagePathAsync(string id, string url)
        {
            // // AkonDeV 06/2026
            string extension = Path.GetExtension(url);
            if (string.IsNullOrEmpty(extension)) extension = ".jpg";
            string cachedPath = Path.Combine(_cacheFolder, $"{id}{extension}");

            if (File.Exists(cachedPath))
            {
                return cachedPath;
            }

            try
            {
                var client = _clientFactory.CreateClient("WallhavenClient");
                byte[] data = await client.GetByteArrayAsync(url);
                await File.WriteAllBytesAsync(cachedPath, data);
                return cachedPath;
            }
            catch (Exception ex)
            {
                Serilog.Log.Error(ex, "Error al descargar imagen a la caché local: {Url}", url);
                return url; // Retorna la URL original como respaldo
            }
        }

        public void ClearCache()
        {
            // // AkonDeV 06/2026
            try
            {
                if (Directory.Exists(_cacheFolder))
                {
                    foreach (var file in Directory.GetFiles(_cacheFolder))
                    {
                        File.Delete(file);
                    }
                }
            }
            catch (Exception ex)
            {
                Serilog.Log.Error(ex, "Error al vaciar la caché.");
            }
        }
    }
}
